service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java

```diff
@@ -185,9 +202,40 @@ public PipelineRun getPipelineRun(UUID jobId, String userId) {
    * status column here. It is currently possible to mark an incomplete pipeline_run as is_success =
    * True using this method.
    */
-  public PipelineRun markPipelineRunSuccess(UUID jobId, String userId) {
+  public PipelineRun markPipelineRunSuccessAndWriteOutputs(
+      UUID jobId, String userId, Map<String, String> outputs) {
     PipelineRun pipelineRun = getPipelineRun(jobId, userId);
     pipelineRun.setIsSuccess(true);
+    pipelineRun.setOutput(outputs);
+
     return pipelineRunsRepository.save(pipelineRun);
   }
+
+  /**
+   * Extract the pipeline outputs from a pipelineRun object, fetch SAS tokens for (currently all of)
+   * them, and return an ApiPipelineRunOutput object with the formatted outputs.
+   *
+   * @param pipelineRun object from the pipelineRunsRepository
+   * @return ApiPipelineRunOutput
+   */
+  public ApiPipelineRunOutput formatPipelineRunOutputs(
+      PipelineRun pipelineRun, String accessToken) {
```

Comment:
i would probably call for access token here unless we're going to be using the token multiple times from where you are calling it in the controller.  This is part of the "business logic" to me so i think it should go in the service layer